### ISSUE TITLE
Additional register with same const behaviour 

### ISSUE THREAD

**Issue by botika** [rust-lang/rust#74020]
In `rustc 1.44.1 (c7087fe00 2020-06-17)` and `linux x86_64`. When use a struct use additional useless register. An example is a simple copy bytes. Without structure https://github.com/botika/buf-min/blob/3ad9833bee4f748f2089ad39fcf9047f9a39b064/benches/src/main.rs#L74-L108 . 
```asm
benches::raw_dyn:
 push    rbx
 sub     rsp, 32
 mov     rbx, rdi
; ...
```
Only push `rbx` register.
While with the structure and exactly the same behavior: https://github.com/botika/buf-min/blob/3ad9833bee4f748f2089ad39fcf9047f9a39b064/benches/src/main.rs#L110-L156
```asm
benches::ibuffer:
 push    r14
 push    rbx
 sub     rsp, 40
 mov     rbx, rdi
 mov     r14d, 12
; ...
```
 An `r14d` I have a constant length, of `"Hello World!"`. 
Therefore, the optimization does not work correctly when using a structure.

**Comment by nagisa** [comment]
>  Therefore, the optimization does not work correctly when using a structure.

Very pedantically speaking the only correctness invariant for optimizers is to produce, as an output, code that is equivalent in behaviour (within some rule-set) to the input code.

**Comment by botika** [comment]
Thanks for answering so quickly. I can look to adjust it and do the PR. It is not a priority. Any suggestions before I get to work on this?

**Comment by nagisa** [comment]
As far as suggestions go, this is very likely to be an issue within LLVM, possibly related to https://github.com/rust-lang/rust/issues/73701. I would bisect rustc nightly versions with your code and see if this is a regression and whether the regression coincides with a LLVM upgrade. In case it is, I would move on to verify that this is not a duplicate of #73701.

**Comment by botika** [comment]
I checked in `rustc 1.39.0 (4560ea788 2019-11-04)` with the same behavior. 
In my case, the conditional and the panic remove them correctly, so it detects that the variables will have a constant value within the structure. 

The problem is that it does not detect that removing the conditional. But it want to build the entire structure even if it don't use it. So use the `r14` register to initialize `xmmword [rsp, +, 8]`. 

If it works with the conditional, `if`, but after that optimization it don't realize what it don't need anymore. In this case you don't need the `len` and `cap` fields of the structure. But still keep the instructions to initialize them.

In other words, it always initializes the full structure.

**Comment by botika** [comment]
I have already looked at nightly, beta, stable and up to 1.39.0 with the same result. 

It doesn't look like a duplicate. 

Register `r14` is used to initialize `xmmword [rsp, +, 8]`. Detects constant behavior for conditional, `if-else`,(`if-else` and `panic` is removed) but it also initializes the whole structure even if it don't need it.

In other words, it always initializes the structures. 

Surely this will be normal LLVM behavior, but it would be nice if it could be done since much of the logic is linked to a structure.
